clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java

```diff
@@ -136,11 +139,55 @@ public synchronized Sensor sensor(String name, MetricConfig config, Sensor... pa
         if (s == null) {
             s = new Sensor(this, name, parents, config == null ? this.config : config, time);
             this.sensors.put(name, s);
+            if (parents != null) {
+                for (Sensor parent : parents) {
+                    List<Sensor> children = childrenSensors.get(parent.name());
+                    if (children == null) {
+                        children = new ArrayList<>();
+                        childrenSensors.put(parent.name(), children);
+                    }
+                    children.add(s);
+                }
+            }
         }
         return s;
     }
 
     /**
+     * Remove a sensor (if it exists), associated metrics and its children.
+     *
+     * @param name The name of the sensor to be removed
+     */
+    public void removeSensor(String name) {
+        List<Sensor> removedSensors = new ArrayList<>();
+
+        synchronized (this) {
+            removeSensor(name, removedSensors);
+        }
+
+        /*
+         * Because `Sensor.metrics` acquires a lock on `sensor` and `Sensor.add` calls methods on `Metrics` that
+         * acquire a lock on the metrics instance, it is important to do the `sensor.metrics` calls below outside
+         * the `synchronized(this)` above.
+         */
+        for (Sensor sensor : removedSensors)
+            for (KafkaMetric metric : sensor.metrics())
+                removeMetric(metric.metricName());
+    }
+
```

Comment:
That may work. I need to go through the code again to double-check. @junrao, would you be happy with this solution instead? If so, I can do it before we merge as it's not a big change.